Client/Frontend/Browser/Search/BraveSearchManager.swift

```diff
@@ -0,0 +1,293 @@
+    
+    /// Brave Search query details which are passed to the fallback provider.
+    struct BackupQuery: Codable {
+        let found: Bool
+        let country: String?
+        let language: String?
+        let safesearch: String?
+    }
+    
+    /// URL of the Brave Search request we performed.
+    /// Contains a query and endpoint(prod or staging)
+    private let url: URL
+    /// What did we search for using Brave Search
+    private let query: String
+    /// BraveSearch cookies are used to pass search settings we saved on the website
+    /// such as whether to use search fallback, should safe search be performed etc.
+    private let domainCookies: [HTTPCookie]
+    
+    /// The result we got from querying the fallback search engine.
+    var fallbackQueryResult: String?
+    /// Whether the call to the fallback search engine is pending.
+    /// This is used to determine at what point of the web navigation we should inject the results.
+    var fallbackQueryResultsPending = false
+    
+    private var cancellables: Set<AnyCancellable> = []
+    private static var cachedCredentials: URLCredential?
+    
+    static func isValidURL(_ url: URL) -> Bool {
+        let validURLs = AppConstants.buildChannel.isPublic ?
+            ["search.brave.com"] : ["search.brave.com", "search-dev.brave.com"]
+        
+        return validURLs.contains(url.host ?? "")
+    }
+    
+    init?(url: URL, cookies: [HTTPCookie]) {
+        if !Self.isValidURL(url) {
+            return nil
+        }
+        
+        // Check if the request has a valid query string to check against.
+        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
+              let queryItem = components.valueForQuery("q") else { return nil }
+        
+        self.url = url
+        self.query = queryItem
+        self.domainCookies = cookies.filter { $0.domain == url.host }
+        if domainCookies.first(where: { $0.name == "fallback" })?.value != "1" {
+            return nil
+        }
+    }
+    
+    /// A call is made to the Brave Search api with query we performed.
+    /// It returns  details that are used in call to the fallback search engine.
+    func shouldUseFallback(completion: @escaping (BackupQuery?) -> Void) {
+        guard var canAnswerURLComponents =
+                URLComponents(string: url.domainURL.absoluteString) else {
+            completion(nil)
+            return
+        }
+        
+        canAnswerURLComponents.scheme = "https"
+        canAnswerURLComponents.path = "/api/can_answer"
+        canAnswerURLComponents.queryItems = [.init(name: "q", value: query)]
+        
+        guard let url = canAnswerURLComponents.url else {
+            completion(nil)
+            return
+        }
+        
+        var request = URLRequest(url: url,
+                                 cachePolicy: .reloadIgnoringLocalAndRemoteCacheData,
+                                 timeoutInterval: 5)
```

Comment:
Is 5 seconds enough? Even for slow networks when using the `extremely bad network` link conditioner?

Reply:
I want it even lower tbh, the website is blocked(white screen) until we return any result, what do you think?